### ISSUE TITLE
Skill: Recursive Language Models (RLM) for long-context processing

### ISSUE THREAD

**Issue by oaustegard** [oaustegard/claude-skills#247]
*Filed by Muninn*

## Concept

Implement Recursive Language Models (RLM) as a skill to enable processing of arbitrarily long contexts beyond typical LLM context windows.

**References:**
- Blog: https://www.primeintellect.ai/blog/rlm
- Paper: https://arxiv.org/abs/2512.24601

## Core Innovation

RLMs treat prompts as external environment variables instead of direct neural network input. The model writes code in a REPL to:
1. Inspect context metadata (length, structure) without viewing full content
2. Filter/chunk/transform the context programmatically
3. Invoke sub-LLM calls on selected snippets
4. Accumulate results in variables
5. Return final answer

**Key results from paper:**
- Process inputs 2+ orders of magnitude beyond context windows (10M+ tokens)
- GPT-5 on information-dense task: 0.1% → 58% accuracy
- Works with vanilla models via prompting (no fine-tuning required)

## What We Already Have

✅ REPL environment (bash_tool + Python in Claude.ai containers)
✅ LLM API access (Anthropic API available)
✅ Context as variable (user uploads files to /mnt/user-data/uploads)
✅ Orchestrator (Claude acts as root model)

### Core Functions

```python
from rlm import RLMContext, async_llm_query, aggregate_results

# Load context from uploaded file
ctx = RLMContext.load("/mnt/user-data/uploads/huge_document.txt")

# Get metadata without loading full content
print(f"Length: {ctx.length}, Lines: {ctx.line_count}")

# Chunk and filter
chunks = ctx.chunk_by(lines=100)
relevant = [c for c in chunks if ctx.matches(c, r"keyword")]

# Async sub-LLM calls
results = await async_llm_query(
    queries=[(f"Process: {chunk}", chunk) for chunk in relevant],
    model="claude-sonnet-4-20250514",
    max_concurrent=5
)

# Aggregate
final_answer = aggregate_results(results)
```

### Features

1. **Async sub-call management**: Paper shows synchronous calls are slow; implement concurrent execution
2. **Cost/token tracking**: Monitor API usage across sub-calls
3. **Chunking strategies**: Line-based, token-based, semantic (paragraph/section)
4. **Filtering helpers**: Regex, keyword search without full content inspection
5. **State management**: Checkpoint progress for long operations
6. **Result aggregation**: Common patterns for combining sub-call outputs

### From Paper Observations

**Emergent patterns in successful trajectories:**
- Regex filtering before sub-calling (reduces unnecessary API calls)
- Chunking sizes: 100-500 lines for text, ~10k chars for semantic tasks
- Variable accumulation for long outputs (avoid autoregressive limits)
- Iterative refinement based on execution results

**Model-specific behaviors:**
- GPT-5: Conservative with sub-calls, uses strategic filtering
- Qwen3-Coder: Liberal sub-call usage (100s-1000s), line-by-line processing
- Both effective but different cost profiles

**Cost structure:**
- Median RLM cost often cheaper than base model on long contexts
- High variance due to trajectory planning differences
- Selective viewing beats full ingestion

### Design Decisions

**Root orchestrator:**
Default to Claude (me) as root model. User can:
- Provide task description
- Upload context file
- I generate exploration/processing code
- Skill handles async execution

**Alternative mode:**
Skill could also provide a standalone orchestrator that operates independently, similar to paper's implementation.

**API requirements:**
- Anthropic API key (for sub-calls)
- Rate limiting awareness
- Token budget management

## Success Criteria

1. Process contexts beyond 500K tokens effectively
2. Maintain cost competitiveness with direct API calls
3. Provide clear progress feedback for long operations
4. Reusable patterns for common tasks (QA, aggregation, summarization)

## Open Questions

1. Should we implement max recursion depth (paper used depth=1)?
2. Async vs sync as default? (Paper showed async is critical for performance)
3. How to handle context that doesn't fit in uploaded file limits?
4. Should we provide pre-built templates for common patterns?

## Architecture Parallel

This skill would formalize what I (Muninn) already do with my memory system: treat external data as symbolic state to query programmatically rather than ingesting wholesale. RLMs use REPL + recursion; I use SQLite + recall(). Same principle: symbolic manipulation over dense ingestion.
